module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,378 @@
+func (s *ExecutionDataService) receiveBatch(br *blobs.BlobReceiver) ([]blobs.Blob, error) {
+	var batch []blobs.Blob
+	var err error
+
+	for i := 0; i < defaultBlobBatchSize; i++ {
+		var blob blobs.Blob
+
+		blob, err = br.Receive()
+
+		if err != nil {
+			break
+		}
+
+		batch = append(batch, blob)
+	}
+
+	return batch, err
+}
+
+// storeBlobs receives blobs from the given BlobReceiver, and stores them to the blobservice
+func (s *ExecutionDataService) storeBlobs(parent context.Context, br *blobs.BlobReceiver, logger zerolog.Logger) ([]cid.Cid, error) {
+	defer br.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	var cids []cid.Cid
+
+	for {
+		batch, recvErr := s.receiveBatch(br) // retrieve one batch at a time
+		batchCids := zerolog.Arr()
+
+		for _, blob := range batch {
+			cids = append(cids, blob.Cid())
+			batchCids.Str(blob.Cid().String())
+		}
+
+		batchLogger := logger.With().Array("cids", batchCids).Logger()
+
+		if err := s.blobService.AddBlobs(ctx, batch); err != nil {
+			batchLogger.Debug().Err(err).Msg("failed to add batch to blobservice")
+
+			return nil, err
+		}
+
+		batchLogger.Debug().Msg("added batch to blobservice")
+
+		if recvErr != nil {
+			if recvErr != blobs.ErrClosedBlobChannel {
+				// this is an unexpected error, and should never occur
+				return nil, recvErr
+			}
+
+			// the blob channel was closed, meaning that all blobs have been received
+			break
+		}
+	}
+
+	return cids, nil
+}
+
+// addBlobs serializes the given object, splits the serialized data into blobs, and adds them to the blobservice
+//
+// blobs are added in a batched streaming fashion, using a separate goroutine to serialize the object and send the
+// blobs of serialized data over a blob channel to the main routine, which adds them in batches to the blobservice.
+func (s *ExecutionDataService) addBlobs(ctx context.Context, v interface{}, logger zerolog.Logger) ([]cid.Cid, error) {
+	bcw, br := blobs.IncomingBlobChannel(s.maxBlobSize)
+
+	done := make(chan struct{})
+	var serializeErr error
+
+	// start serialization goroutine
+	go func() {
+		defer close(done)
+		defer bcw.Close()
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			logger.Debug().Err(serializeErr).Msg("failed to serialize execution data")
+
+			return
+		}
+
+		serializeErr = bcw.Flush()
+
+		if serializeErr != nil {
+			logger.Debug().Err(serializeErr).Msg("flushed execution data")
+		}
+	}()
+
+	cids, storeErr := s.storeBlobs(ctx, br, logger)
+
+	<-done
+
+	if storeErr != nil {
+		return nil, storeErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *ExecutionDataService) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	logger := s.logger.With().Str("block_id", sd.BlockID.String()).Logger()
+	logger.Debug().Msg("adding execution data")
```

Comment:
better to add an info level log with both block ID, cids as well as the root CID, so that their relations are logged together.